File: sdks/risk-client-go/channel.go
package riskclient

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/coder/websocket"
	"github.com/coder/websocket/wsjson"
)

// ChannelClient creates WebSocket connections.
type ChannelClient struct {
	wsURL  string
	apiKey string
}

func newChannelClient(cfg Config) *ChannelClient {
	base := envMap[cfg.Environment].restBaseURL
	base = strings.ReplaceAll(base, "https://", "wss://")
	base = strings.ReplaceAll(base, "http://", "ws://")
	return &ChannelClient{wsURL: base + "/ws/risk", apiKey: cfg.APIKey}
}

// RiskChannel is a live bidirectional WebSocket connection.
type RiskChannel struct {
	conn *websocket.Conn
}

// Open dials the WebSocket endpoint and returns a ready channel.
func (c *ChannelClient) Open(ctx context.Context) (*RiskChannel, error) {
	conn, _, err := websocket.Dial(ctx, c.wsURL, &websocket.DialOptions{
		HTTPHeader: map[string][]string{"X-API-Key": {c.apiKey}},
	})
	if err != nil {
		return nil, &RiskClientError{Message: fmt.Sprintf("WebSocket dial %s", c.wsURL), Cause: err}
	}
	return &RiskChannel{conn: conn}, nil
}

// Send serializes req and writes it to the WebSocket.
func (ch *RiskChannel) Send(ctx context.Context, req RiskRequest) error {
	return wsjson.Write(ctx, ch.conn, req)
}

// Receive reads one message and deserializes it as a RiskDecision.
func (ch *RiskChannel) Receive(ctx context.Context, timeoutMs int) (*RiskDecision, error) {
	readCtx, cancel := context.WithTimeout(ctx, time.Duration(timeoutMs)*time.Millisecond)
	defer cancel()

	var raw json.RawMessage
	if err := wsjson.Read(readCtx, ch.conn, &raw); err != nil {
		return nil, err
	}
	var dec RiskDecision
	if err := json.Unmarshal(raw, &dec); err != nil {
		return nil, err
	}
	return &dec, nil
}

// Close shuts down the WebSocket connection gracefully.
func (ch *RiskChannel) Close() {
	_ = ch.conn.Close(websocket.StatusNormalClosure, "client close")
}

File: sdks/risk-client-typescript/src/channel.ts
import WebSocket from 'ws';
import { ClientOptions, RiskDecision, RiskRequest } from './types';
import { resolveEnv } from './env';

export class RiskChannel {
  private readonly inbox: RiskDecision[] = [];
  private readonly waiters: Array<(d: RiskDecision) => void> = [];

  constructor(private readonly ws: WebSocket) {
    ws.on('message', (data) => {
      try {
        const decision = JSON.parse(data.toString()) as RiskDecision;
        const waiter = this.waiters.shift();
        if (waiter) {
          waiter(decision);
        } else {
          this.inbox.push(decision);
        }
      } catch { /* skip */ }
    });
  }

  send(req: RiskRequest): Promise<void> {
    return new Promise((resolve, reject) => {
      this.ws.send(JSON.stringify(req), (err) => {
        if (err) reject(err); else resolve();
      });
    });
  }

  receive(timeoutMs = 2000): Promise<RiskDecision | null> {
    if (this.inbox.length > 0) {
      return Promise.resolve(this.inbox.shift()!);
    }
    return new Promise((resolve) => {
      const timer = setTimeout(() => {
        const idx = this.waiters.indexOf(resolve);
        if (idx !== -1) this.waiters.splice(idx, 1);
        resolve(null);
      }, timeoutMs);

      this.waiters.push((d) => {
        clearTimeout(timer);
        resolve(d);
      });
    });
  }

  close(): void {
    this.ws.close(1000, 'client close');
  }
}

export class ChannelFactory {
  private readonly wsUrl: string;

  constructor(private readonly options: ClientOptions) {
    const base = resolveEnv(options.environment).restBaseUrl
      .replace('https://', 'wss://')
      .replace('http://', 'ws://');
    this.wsUrl = `${base}/ws/risk`;
  }

  open(): Promise<RiskChannel> {
    return new Promise((resolve, reject) => {
      const ws = new WebSocket(this.wsUrl, {
        headers: { 'X-API-Key': this.options.apiKey },
      });
      ws.once('open', () => resolve(new RiskChannel(ws)));
      ws.once('error', reject);
    });
  }
}
